app/components/select-tree.tsx

```diff
@@ -0,0 +1,351 @@
+  label,
+  options,
+  value,
+  onChange,
+  tooltipText,
+  disabled,
+  controls,
+}: SelectTreeProps) {
+  const [anchorEl, setAnchorEl] = useState<HTMLElement>();
+  const [minMenuWidth, setMinMenuWidth] = useState<number>();
+
+  const handleClick = useEventCallback((ev: React.MouseEvent<HTMLElement>) => {
+    setAnchorEl(ev.currentTarget);
+    setMinMenuWidth(ev.currentTarget.clientWidth);
+  });
+  const handleClose = useEventCallback(() => {
+    setAnchorEl(undefined);
+  });
+  const open = !!anchorEl;
+  const classes = useStyles({ disabled, open });
+
+  const parentsRef = React.useRef({} as Record<NodeId, NodeId>);
+  const labelsByValue = useMemo(() => {
+    parentsRef.current = {} as Record<NodeId, NodeId>;
+    const res: Record<string, string> = {};
+    const registerNode = ({ value, label, children }: Tree[number]) => {
+      res[value] = label;
+      if (children && children.length > 0) {
+        for (let child of children) {
+          registerNode(child);
+          parentsRef.current[child.value as unknown as NodeId] = value;
+        }
+      }
+    };
+    for (let root of options) {
+      registerNode(root);
+    }
+    return res;
+  }, [options]);
+
+  const defaultExpanded = useMemo(() => {
+    const res = [];
+    let cur = value;
+    const parents = parentsRef.current;
+    while (cur && parents[cur]) {
+      res.push(parents[cur]);
+      cur = parents[cur];
+    }
+    return res;
+  }, [value]);
+
+  const handleNodeSelect = useEventCallback((_ev, value: NodeId) => {
+    onChange({ target: { value: value } });
+    handleClose();
+  });
+
+  const treeItemClasses = useTreeItemStyles();
+  const renderTreeContent = useCallback(
+    (nodesData: Tree) => {
+      return (
+        <>
+          {nodesData.map(({ value, label, children, selectable }) => {
+            return (
+              <TreeItem
+                key={value}
+                nodeId={value}
+                defaultExpanded={defaultExpanded}
+                label={label}
+                expandIcon={
+                  children && children.length > 0 ? <ChevronRightIcon /> : null
+                }
+                classes={treeItemClasses}
+                ContentProps={{
+                  // @ts-expect-error - TS says we cannot put a data attribute
+                  // on the HTML element, but we know we can.
+                  "data-selectable": selectable,
+                }}
+              >
+                {children ? renderTreeContent(children) : null}
+              </TreeItem>
+            );
+          })}
+        </>
+      );
+    },
+    [defaultExpanded, treeItemClasses]
+  );
+
+  const menuRef = React.useRef<PopoverActions>(null);
+  const handleNodeToggle = () => {
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 500);
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 1000);
+    setTimeout(() => {
+      menuRef.current?.updatePosition();
+    }, 2000);
+  };
```

Comment:
Why such combo? Wouldn't just first timeout do the job? Or is the transition duration dependent on the number of children in a node?

Reply:
Refactored away by using the onEntered / onExited transition props of the tree item transitions.